ui/src/main/webapp/src/app/components/reports/source/source-report.component.ts

```diff
@@ -0,0 +1,283 @@
+                .subscribe((classifications) => this.classifications = classifications );
+
+            this.hintService.getHintsForFile(this.execID, this.fileID)
+                .subscribe((hints) => this.hints = hints);
+
+            this.fileModelService.getSource(this.execID, this.fileID)
+                .subscribe((fileSource) => {
+                    this.fileSource = fileSource;
+                    this.fileLines = fileSource.split(/[\r\n]/).map((line) => line + "\n");
+                });
+        });
+    }
+
+    private getClassificationLinks(classification:ClassificationModel):Observable<LinkModel[]> {
+        let linkableModel = <LinkableModel>new GraphJSONToModelService().translateType(classification, this.http, LinkableModel);
+        return linkableModel.links;
+    }
+
+    get storyPoints():number {
+        let points = 0;
+        if (this.hints)
+            this.hints.forEach((hint) => points += hint.effort);
+
+        if (this.classifications)
+            this.classifications.forEach((classification) => points += classification.effort);
+
+        return points;
+    }
+
+    private hintMatches(hint:InlineHintModel, lineNumber:number):boolean {
+        let hintLine = hint.data["lineNumber"];
+
+        // workaround an odd edge case
+        if (hintLine <= 0)
+            hintLine = 1;
+
+        return hintLine == (lineNumber+1)
+    }
+
+    noteReferences(line:string, lineNumber:number): string {
+        if (!this.hints)
+            return "";
+
+        return this.hints
+            .filter((hint) => this.hintMatches(hint, lineNumber))
+            .map((hint) => "note-" + hint.vertexId)
+            .join(", ");
+    }
+
+    lineClass(line:string, lineNumber:number): string {
+        if (!this.hints)
+            return "";
+
+        let styleClasses = this.hints
+            .filter((hint) => this.hintMatches(hint, lineNumber))
+            .map((hint) => "note-" + hint.vertexId)
+            .join(", ");
+
+        if (styleClasses)
+            styleClasses = "note-placeholder box box-bg " + styleClasses;
+
+        return styleClasses;
+    }
+
+    filetype():string {
+        if (!this.fileModel)
+            return "";
+
+        let lastDotIndex = this.fileModel.fileName.lastIndexOf(".");
+        if (lastDotIndex == -1 || lastDotIndex == this.fileModel.fileName.length)
+            return "";
+
+        return this.fileModel.fileName.substring(lastDotIndex + 1);
+    }
+
+    markdown(input:string):string {
+        return new showdown.Converter().makeHtml(input);
+    }
+
+    ngAfterViewChecked(): void {
+        if (this.rendered)
+            return;
+
+        if (!this.fileSource || !this.fileModel || this.hints == null || this.classifications == null)
+            return;
+
+        Prism.hooks.add('after-highlight', function () {
+            [].forEach.call(document.querySelectorAll('.has-notes .note-placeholder'), function (placeholder) {
```

Comment:
Oh, I see, this is a hack because NodeListOf<Element> does not have a forEach function. The code works around this by using the forEach method from [], which actually works. I think this is a really ugly way to do it, though.